third_party/move/move-compiler-v2/src/bytecode_generator.rs

```diff
@@ -791,12 +791,14 @@ impl<'env> Generator<'env> {
             Operation::Shr => self.gen_op_call(targets, id, BytecodeOperation::Shr, args),
             Operation::And => self.gen_logical_shortcut(true, targets, id, args),
             Operation::Or => self.gen_logical_shortcut(false, targets, id, args),
-            Operation::Eq => self.gen_op_call(targets, id, BytecodeOperation::Eq, args),
-            Operation::Neq => self.gen_op_call(targets, id, BytecodeOperation::Neq, args),
-            Operation::Lt => self.gen_op_call(targets, id, BytecodeOperation::Lt, args),
-            Operation::Gt => self.gen_op_call(targets, id, BytecodeOperation::Gt, args),
-            Operation::Le => self.gen_op_call(targets, id, BytecodeOperation::Le, args),
-            Operation::Ge => self.gen_op_call(targets, id, BytecodeOperation::Ge, args),
+            Operation::Eq => self.gen_op_call_auto_freeze(targets, id, BytecodeOperation::Eq, args),
+            Operation::Neq => {
+                self.gen_op_call_auto_freeze(targets, id, BytecodeOperation::Neq, args)
+            },
+            Operation::Lt => self.gen_op_call_auto_freeze(targets, id, BytecodeOperation::Lt, args),
```

Comment:
Why do we need "auto freeze" for `Lt`, `Gt`, `Ge`, `Le`? I don't think references even be arguments to these operations.

I understand that we need it for `==` and `!=`, based on the description here: https://aptos.dev/en/build/smart-contracts/book/equality#typing-with-references.

Reply:
They might be in the future, once we introduce comparison as a general operator on arbitrary types. Here I anticipated this to avoid subtle errors once this change happens.

Reply:
That still isn't a good argument.  Unless we want to allow pointer arithmetic and deterministic field ordering in objects, we have no reason to ever allow anything except `==` and `!=` on references.

But, it probably doesn't hurt anything, so fine.

Reply:
I'm not quite following. In Rust all comparison operators work on references via the basic `Cmp` trait.